lighthouse-core/config/v2/config.js

```diff
@@ -0,0 +1,186 @@
+    configJson = ConfigV2.extendIfNecessary(configJson, configPath);
+
+    this._json = configJson;
+    this._gatherers = ConfigV2.collectImplementations(configJson.gatherers);
+    this._audits = ConfigV2.collectImplementations(configJson.audits);
+    this._passes = ConfigV2.computePasses(configJson, this._gatherers, this._audits);
+    this._report = configJson.report;
+  }
+
+  asJson() {
+    return JSON.parse(JSON.stringify(this._json));
+  }
+
+  get audits() {
+    return [...this._audits];
+  }
+
+  get passes() {
+    return [...this._passes];
+  }
+
+  get report() {
+    return this._report;
+  }
+
+  static _tryResolveUntilSuccess(paths) {
+    for (const path of paths) {
+      try {
+        return require.resolve(path);
+      } catch (e) {}
+    }
+
+    throw new Error(`Unable to locate ${paths[0]}`);
+  }
+
+  static _mergeObjects(base, extension) {
+    if (!base ||
+        !extension ||
+        typeof extension !== 'object' ||
+        typeof base !== 'object' ||
+        Array.isArray(base)) {
+      return extension;
+    }
+
+    for (const key in extension) {
+      if (key !== 'extends') {
+        base[key] = ConfigV2._mergeObjects(base[key], extension[key]);
+      }
+    }
+
+    return base;
+  }
+
+  static extendIfNecessary(configJson, configPath) {
+    if (!configJson.extends) {
+      return configJson;
+    }
+
+    let extendedConfigJson;
+    let extendedConfigPath = configJson.extends;
+    if (extendedConfigPath === 'lighthouse:default') {
+      extendedConfigJson = defaultConfigJson;
+      extendedConfigPath = defaultConfigPath;
+    } else {
+      extendedConfigPath = ConfigV2._tryResolveUntilSuccess([
+        extendedConfigPath,
+        path.resolve(configPath, extendedConfigPath),
+        path.resolve(process.cwd(), extendedConfigPath),
+      ])
+      extendedConfigJson = require(extendedConfigPath);
+    }
+
+    const extendedConfig = new ConfigV2(extendedConfigJson, extendedConfigPath);
+    return ConfigV2._mergeObjects(extendedConfig.asJson(), configJson);
+  }
+
+  static resolvePaths(object, configPath, searchPaths = []) {
+    object = Object.assign({}, object);
+    Object.keys(object).forEach(key => {
+      // We don't need to resolve objects that already have
+      if (object[key].implementation) {
+        return;
+      }
+
+      const rawPath = object[key].path || key;
+      const possiblePaths = searchPaths.map(searchPath => {
+        return path.resolve(searchPath, rawPath);
+      }).concat([
+        rawPath, // for npm plugins and absolute path usage
+        path.resolve(configPath, rawPath), // for relative config usage
+        path.resolve(process.cwd(), rawPath), // for node module usage
+      ]);
+      const resolvedPath = ConfigV2._tryResolveUntilSuccess(possiblePaths);
+      object[key] = Object.assign({}, object[key], {path: resolvedPath});
+    });
+    return object;
+  }
+
+  static objectToList(object) {
```

Comment:
mainly for testing since it did do fancier things like ordering, but that's been nixed for a future PR and didn't seem worth moving